core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,351 @@
+                                                      .labelNames(labels)
+
+                                                      .help(namingFunction.getDescription())
+                                                      .create());
+
+            }
+            return this;
+        }
+
+        private <T extends Collector> T create(String name,
+                                               PrometheusRegistryWrapper collectorRegistry,
+                                               Function<String, T> ifAbsent) {
+            return collectorRegistry.create(name, ifAbsent);
+        }
+
+        private <T> String[] getValues(MetricLabel<T> label) {
+            BiFunction<Method, T, Object> callMethod = (method, obj) -> {
+                try {
+                    return method.invoke(obj);
+                } catch (Throwable t) {
+                    throw new RuntimeException(t);
+                }
+            };
+
+            return metricLabelValueMethods.stream()
+                                          .map(method -> callMethod.apply(method, label.getValue()))
+                                          .map(Objects::toString)
+                                          .toArray(String[]::new);
+        }
+
+        public MetricAdapter time(MetricLabel<?> label, long time) throws Exception {
+            check(timer).labels(getValues(label)).observe(time);
+            return this;
+        }
+
+        public MetricAdapter success(MetricLabel<?> label) {
+            check(success).labels(getValues(label)).inc();
+            return this;
+        }
+
+        public MetricAdapter failure(MetricLabel<?> label) {
+            check(failure).labels(getValues(label)).inc();
+            return this;
+        }
+
+        public MetricAdapter incActiveRequests(MetricLabel<?> label) {
+            check(activeRequests).labels(getValues(label)).inc();
+            return this;
+        }
+
+        public MetricAdapter decActiveRequests(MetricLabel<?> label) {
+            check(activeRequests).labels(getValues(label)).dec();
+            return this;
+        }
+
+        public MetricAdapter requestBytes(MetricLabel<?> label, RequestLog log) {
+            check(requestBytes).labels(getValues(label)).observe(log.requestLength());
+            return this;
+        }
+
+        public MetricAdapter responseBytes(MetricLabel<?> label, RequestLog log) {
+            check(responseBytes).labels(getValues(label)).observe(log.responseLength());
+            return this;
+        }
+
+        private <T> T check(T metric) {
+            if (Objects.isNull(metric)) {
+                throw new IllegalStateException(MetricAdapter.class.getCanonicalName() + "not initialized");
+            }
+            return metric;
+        }
+    }
+
+    private PrometheusRegistryWrapper collectorRegistry;
+    private Function<RequestLog, MetricLabel> labelingFunction;
+    private MetricAdapter metricAdapter;
```

Comment:
The semantics seem pretty weird to me - this decorator, so this metricAdapter, are basically singletons, but the adapter is initialized lazily on the first request, and we don't even know what type of request it is. So there's an implicit requirement that the labeling function always returns all possible names - it'd be better to make it explicit and remove the lazy initialization, along with its complexity here.

The bean is also sort of magical - I guess all the properties have to return something with a reasonable toString representation. Are you sure you want to be able to have beans returning e.g., a List?

Instead of a bean, how about just using a Map of Labels which is more type-safe and makes sure users control the actual string representation they want?

```java
interface Label {
  String name();
}

Function<RequestLog, Map<? extends Label, String>> labelingFunction
```

And an enum would be convenient for general usage
```java
enum RequestLabels implements Label {
  os,
  version,
  blah
}
```

The decorator constructor could then accept `Label[]` and know all the names before hand and not need to do lazy initialization

```java
newDecorator(RequestLabels.values());
```

The initialized guard goes away, check goes away, reflection goes away, fields become final - lots of goodness IMO :)